types.go

```diff
@@ -125,7 +126,10 @@ func NewMeteredOptionalType(gauge common.MemoryGauge, typ Type) OptionalType {
 func (OptionalType) isType() {}
 
 func (t OptionalType) ID() string {
-	return fmt.Sprintf("%s?", t.Type.ID())
+	if len(t.typeID) == 0 {
+		t.typeID = fmt.Sprintf("%s?", t.Type.ID())
```

Comment:
Here, and for other non-pointer receiver `ID` functions: The receiver is mutated here, but because it is not a pointer, this mutation is "lost" after the function returns.

To effectively cache IDs, we'll likely need to turn all current non-pointer types conforming to `cadence.Type` to pointer types.

Reply:
Good catch, added with https://github.com/onflow/cadence/pull/2336